### PR TITLE
Unimpliment coalesce from spark script and make necessary changes to spark runner

### PR DESCRIPTION
**Pull request by Sami1309**
# Description

Removes the .coalesce option called when saving resources in the spark script. This saves the resources in multiple parts instead of conjoining them into a single part. Implementing this requires changing how iterators parse through files and the functional definition of a resource in the offline store (so now a resource is defined as the parent directory containing all the parts, not the single parquet file)

### Select type(s) of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

# Checklist:

- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have fixed any merge conflicts

### CODE REVIEW

File: provider/spark.go [hunk 1]
```diff
@@ -783,7 +828,13 @@ func (s *S3PrimaryTable) GetName() string {
 }
 
 func (s *S3PrimaryTable) IterateSegment(n int64) (GenericTableIterator, error) {
-	columnStream, err := s.store.ResourceStreamConv(s.sourcePath, 0)
+	var columnStream chan interface{}
+	var err error
+	if s.isTransformation {
```
Comment (sdreyer) [inline provider/spark.go:833]: Why do we care if its a transformation or not?

File: provider/spark.go [hunk 2]
```diff
@@ -250,8 +250,11 @@ type SparkExecutor interface {
 type SparkStore interface {
 	UploadSparkScript() error //initialization function
 	ResourceKey(id ResourceID) (string, error)
+	ResourceKeys(id ResourceID) ([]string, error)
 	ResourceStreamConv(key string, begin int64) (chan interface{}, error)
-	ResourceRowCt(key string) (int, error)
+	ResourceMultiPartStream(id ResourceID) (chan interface{}, error)
+	ResourceRowCt(id ResourceID) (int, error)
```
Comment (sdreyer) [inline provider/spark.go:256]: Can you just rename these to Count instead of Ct?

File: provider/spark.go [hunk 3]
```diff
@@ -250,8 +250,11 @@ type SparkExecutor interface {
 type SparkStore interface {
 	UploadSparkScript() error //initialization function
 	ResourceKey(id ResourceID) (string, error)
+	ResourceKeys(id ResourceID) ([]string, error)
 	ResourceStreamConv(key string, begin int64) (chan interface{}, error)
-	ResourceRowCt(key string) (int, error)
+	ResourceMultiPartStream(id ResourceID) (chan interface{}, error)
```
Comment (sdreyer) [inline provider/spark.go:255]: This interface is really confusing, but refactoring is out of scope for this. Will create a task to clean this up after so we can get this working

File: provider/spark.go [hunk 4]
```diff
@@ -1119,18 +1177,21 @@ func (s *S3Materialization) ID() MaterializationID {
...
 	return int64(numRows), nil
 }
 
 func (s *S3Materialization) IterateSegment(begin, end int64) (FeatureIterator, error) {
-	stream, err := s.store.ResourceStreamConv(s.Key, begin)
+	stream, err := s.store.ResourceMultiPartStream(s.id)
 	if err != nil {
 		return nil, err
 	}
+	for i := int64(0); i < begin; i++ {
```
Comment (sdreyer) [inline provider/spark.go:1192]: Wasn't there a function that does this?

File: provider/spark.go [hunk 5]
```diff
@@ -356,6 +359,35 @@ func ResourcePrefix(id ResourceID) string {
 	return fmt.Sprintf("featureform/%s/%s/%s/", id.Type, id.Name, id.Variant)
 }
 
+func (s *S3Store) ResourceKeys(id ResourceID) ([]string, error) {
```
Comment (sdreyer) [inline provider/spark.go:362]: What does this mean?

File: provider/spark.go [hunk 6]
```diff
@@ -356,6 +359,35 @@ func ResourcePrefix(id ResourceID) string {
...
+		s.logger.Errorw("Failure retrieving objects from S3 Store", err)
+		return nil, err
+	}
+	if len(objects.Contents) == 0 {
+		s.logger.Errorw("no objects with key prefix exist in S3", latestFilePrefix)
+		return []string{}, nil
+	}
+	s.logger.Debugf("Found %d objects", len(objects.Contents))
+	returnList := make([]string, 0, len(objects.Contents)-1)
+	for _, object := range objects.Contents {
```
Comment (sdreyer) [inline provider/spark.go:382]: Unclear what this is doing

File: provider/spark.go (hunk 4 above)
Comment (Sami1309) [reply provider/spark.go]: Added an implementation for multiple files

**Comment by sdreyer** [comment]
@Sami1309 can you run this with the tests @ahmadnazeri wrote and verify those still work?

**Review by sdreyer — APPROVED** [review]

**Pull request merged**